The Specs behind the Specs (#50)
working-group-two/engineering-blog: pages/blog/_posts/2022-01-10-the-spec-behind-the-specs.md

```
notValidReqInfoType RequestedInformationType ::= 4
```

### BIT STRING

`BIT STRING` takes values that are a sequence of zero or more bits. It
can also take an additional notation that name certain bits in the bit
sequence.

```
DeferredLocationEventType ::= BIT STRING {
    msAvailable (0) ,
    enteringIntoArea (1),
    leavingFromArea (2),
    beingInsideArea (3) ,
    periodicLDR (4)
} (SIZE (1..16))
```

```
eventType1 DeferredLocationEventType ::= (msAvailable, beingInsideArea)
eventType2 DeferredLocationEventType ::= '10010'B
eventType3 DeferredLocationEventType ::= '12'H
```

are all valid value definitions of the same bit sequence where the
first and third bits are set, and no other bits are set.  The `B` stands
for binary representation and `H` for hexadecimal representation.

The `SIZE` is a constraint on the type defining it to be of a specific
length. This keyword comes as an extra notation for many of the
`STRING` types below (as well as some of the other types).

### OCTET STRING

Type `OCTET STRING` takes values that are an ordered sequence of zero
or more (eight-bit) octets.

```
MM-Code ::= OCTET STRING (SIZE (1))
```

> In the same manor as `BIT STRING` both values below are valid

Reviewer: ```suggestion
In the same manner as `BIT STRING` both values below are valid
```

instances of `MM-Code`:

```
iMSI-Attach1 MM-Code ::= '00000010'B
iMSI-Attach2 MM-Code ::= '02'H

> ```

Author: only H and B octet strings are supported. in order to en/decode ascii one would probably need to add some encoding-rules IDK.

Historically some of the strings are defined by using octet strings
`IA5String ::= [UNIVERSAL 22] IMPLICIT OCTET STRING`

but I guess it is up to the compiler to take care of it.

| UniversalString              | 28  | [ISO10646-1]                                                                                                                                                                                  |
| BMPString                    | 30  | Basic Multilingual Plane; subtype of UniversalString                                                                                                                                          |

[ISOReg](https://www.itscj-ipsj.jp/custom_contents/cms/linkfile/ISO-IR.pdf)
is a pretty good source, it reference most of the registers but not
all of them as far as I can see.

I'll list some examples of string types found in our ASN.1 files:

```
AMFNameUTF8String ::= UTF8String (SIZE(1..150, ...))

DirectoryString ::= CHOICE {
    teletexString TeletexString (SIZE (1..maxSize)),
    printableString PrintableString (SIZE (1..maxSize)),
    universalString UniversalString (SIZE (1..maxSize)),
    bmpString BMPString (SIZE (1..maxSize))
--    utf8String UTF8String (SIZE (1..maxSize))
    }

DisplayInformation ::= IA5String (SIZE (minDisplayInformationLength..maxDisplayInformationLength))
```

> `IA5String` used to represent ISO 646 (IA5; International Alphabet 5)

Reviewer: ```suggestion
`IA5String` is used to represent ISO 646 (IA5; International Alphabet 5)
```

  interruptableAnnInd  [8] BOOLEAN DEFAULT TRUE,
  voiceInformation     [9] BOOLEAN DEFAULT FALSE,
  voiceBack            [10] BOOLEAN DEFAULT FALSE
}
```

In this sequence we can see that many of the values are optional and
some have defaults, only the value `maximumNbOfDigits` is mandatory.

When the sending side transmits a value of `CollectedDigits` type, the
receiving side will get a sequence of the values mentioned. With
BER-encoding each defined value will have an identifier, a length (of
the value transmitted) and the value. This is called a
Tag-Length-Value or TLV for short.

All basic types already has an universal tag as stated in the [types
table](#types), but the composite types does not. (If not tagged, how
would it see the difference between `endOfReplyDigit`, `cancelDigit`
and `startDigit` in the example above?)

When explicitly tagged (as above) the both the tag `[2]` and the
universal tag `[4]` for `OCTET STRING` would be transmitted for
`endOfReplyDigits`. This is all fine and well, because then the types
could be potentially extended, the `OCTET STRING` could for instance
be replaced with a `BIT STRING` and the receiver would still

> understand the message because the universal tag `[3]` would be sent

Reviewer: Hm, I don't know if this is realistic. The receiver code was most likely created from the ASN.1 spec which specified an OCTET STRING.

Author: Removed

  forwardedToNumber  [1] IMPLICIT IsdnNumber
}

IsdnNumber ::= SEQUENCE {
  typeOfAddress  TypeOfAddress,
  digits         TelephonyString
}

TypeOfAddress ::= ENUMERATED {national(0), international(1), private(2)}

TelephonyString ::=
  IA5String
    (FROM ("0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" | "*" |
           "#"))(SIZE (1..15))

```

A value of `RoutingInformation` in this example will only transmit the
tag `[0]` or `[1]` (and of course the length and value). It counts on
that the receiving part has the same version of the ASN.1 and knows
the abstract syntax. If it wouldn't have been `IMPLICIT`, then it
would have sent either tags `[0]` or `[1]` followed by the tags for
`TypeOfAddress` (ENUMERATED) `[10]`, and tags for `TelephonyString`
(IA5String) `[4]`.

One can specify `IMPLICIT` and `EXPLICIT` tagging on module basis,
where the `DEFINITIONS` are assigned.

```
DEFINITIONS IMPLICIT TAGS ::= BEGIN
```

Instead of writing all the tags self (explicitly in both cases), one
can instead use the keywords `AUTOMATED TAGS`.

```
DEFINITIONS AUTOMATIC TAGS ::= BEGIN
```

This will add tags to the composit types that doesn't have them
(explicitly) set.

> ## Deprecations and discurragements
> 
> Some things have been deprecated from earlier ASN.1 specifications,
> and use of these are strongly discurraged.

Reviewer: ```suggestion
## Deprecations and discouragements

Some things have been deprecated from earlier ASN.1 specifications,
and use of these are strongly discouraged.
```

NOTATION`. The rest of the fields are the value sequence defining what
the macro should insert. Quotes define the string to insert (excluding
the actual quotes), `empty` inserts nothing. `identifier`, `value`, or
`type` are used to infer different things. Dubuisson has (yet again) a
good chapter on this topic.

Also here one should use informal object classes and parameterized
components instead of using macros.

# Encodings

There are numerous codecs when transmitting the abstract syntax, all
with different pros and cons.

| Short name | Long name                    |
|------------|------------------------------|
| BER        | Binary encoding rules        |
| DER        | Distinguished encoding rules |
| CER        | Canonical encoding rules     |
| PER        | Packed encoding rules        |
| OER        | Octet encoding rules         |
| XER        | XML encoding rules           |
| EXER       | Extended XML encoding rules  |
| JER        | JSON encoding rules          |

BER is the oldest encoding rule for ASN.1. It uses Tag-Length-Value
format.

DER and CER are subsets of BER. CER is mostly used for X.509 digital
certificates.

> PER is the most compact format, and used for bandwith conservation. It

Reviewer: Maybe also mention U-PER and that that is not aligned to octet boundaries

does not send the Tag of the TLV because the order in which components
of the message occur is known.  PER also does not send the Length of
the TLV if the Value has a fixed length. Uses information from ASN.1
message description to eliminate redundant information from the Value
portion.

OER uses an octet oriented format, so the length of all
Tag-Length-Values are padded to be multiples of 8 bits. This makes it

> the fastest ASN.1 encoding and decoding.

Reviewer: Wait, isn't that the same as BER?!
